### PR TITLE
Breach Arbiter Persistence

### PR DESCRIPTION
**Pull request by cfromknecht**
This PR extends the work previously done in PR #201 by @phlip9,  and brings full persistence to the breach arbiter and the retribution process as discussed in issue #184. It also makes general improvements to the stability and reliability of the subsystem as a whole.

Channel States
---------------------
The breach arbiter tracks channels in four primary states as follows:
1. `NB`: Not-Breached -- an open channel for which no breach has been detected.
2. `CB`: Commit-Breached -- a breach has been detected and the arbiter has persisted the incident to disk, however the channel remains open in channeldb.
3. `BC`: Breach-Closed -- the breach is both persisted to by the arbiter and the channel has been closed in channeldb using the CloseBreach enum.
4. `JS`: Justice-Served -- the breach has been resolved via confirmation of the justice transaction.

Failure Modes
---------------------
In order to ensure safe passage from `NB` to `JS`, we must ensure that the arbiter can tolerate failures between any intermediary transition, i.e. `NB -> CB`, `CB -> BC`, and `BC -> JS`.  The breach arbiter will pursue the appropriate recovery measure in response to each event as follows:
1. `NB -> CB`: A failure during this transition means that a breach has been detected by lnd, but has yet to be recorded.  During startup, the breach arbiter will treat the channel as if it were still open and resume monitoring as usual, which includes receiving chain notifications for the active channels. Upon delivery of tx conf notification, the retribution process will resume as if there hadn't been a crash.
2. `CB -> BC`: The breach has been persisted, but the channel remains open. The persisted information encodes all relevant information to deterministically initiate the exact channel close that would have been used if there had been no crash.  During a restart, the breach arbiter checks for channels in this state and proceeds by reinitiating the channel close to channeldb.
3. `BC -> JS`: The breach is persisted and the channel is closed. This transition succeeds upon receiving confirmation of the justice transaction in the chain.  During a restart, the arbiter spawns an `exactRetribution` task for all `BC` channels, which is responsible for attempting the broadcast and waiting for the necessary confirmations.  This will silently fail if justice has already been served.

Restart Details
---------------
During an uninterrupted retribution procedure, the breach arbiter tracks the channel through each of the four states sequentially.  During restart, great care is taken to ensure that we avoid any invalid state transitions. Some general observations regarding the system that this design leverages:
1. Of the three transitions, the `CB -> BC` transition is both initiated by the breach arbiter and free of network calls. It is thus the cheapest transition and can be applied efficiently during restarts.
2. Any `BC` channel has a corresponding entry in the retribution store. This entry can only be removed if an `exactRetribution` task is spawned, eventually removing the entry from the retribution store.

During a restart, the breach arbiter will attempt to recover the retribution process for any `CB` or `BC` channels.  It does so first by promoting all `CB` channels to `BC`, by reinitiating the write to channeldb.  This ensures that _the breach observer will never finish a restart with channels in the `CB` state._ Given observation 1, this can be done very efficiently and simplifies the remaining recovery procedure.

Since we now only need to consider `BC` channels, the remainder of the recovery procedure uses observation 2. All entries marked `BC` in the retribution store will eventually need to have justice delivered in order to be removed! Simply spawn one task to finish out the justice process. Since all of these channels are no longer open according to channeldb, they do not need to be monitored by the contractObserver or a particular breachObserver.

Other Improvements
--------------------
 * Adds test to ensure Alice resumes the road to justice if she crashes just after having received a confirmation for the breach transaction.
 * Generic RetributionStore interface and small test suite for basic functionality, behavior, and persistence requirements.
 * Improved goroutine tracking and cleanup, rounding up a number of previously untracked goroutines.
 * Adds conditional channel sends to ensure all goroutines exit during shutdown. Fixes bug where blocking sends cause lnd to stall during exit because receiver has already quit. Hard to recover if you can't exit 😉

Notes
------
  * It should be noted that in its current state, the entries in the retribution store are not always removed properly.  This is benign as it's okay to attempt justice more than once. @Roasbeef has suggested that well wait to solidify this logic until we've refactoring the errors produced by the wallet's `PublishTransaction` method.
 * The breach arbiter currently does not sweep second layer transactions, this will be included in a separate PR.
* Some refactoring of the witness function generation and test suites are already in progress, and will be merged in a separate PR. This is primarily in attempt to keep this diff from sprawling too much, and to keep the process of reviewing refactored code more isolated from that of reviewing new functionality or behavior.
 * There may still be edge cases related to the fact that an htlc switch can be reopened before justice is fully carried out, meaning that the persisted channel snapshot data may be stale. This should be addressed by check pointing the breach further upstream, including additional logic to restart the delivery of the breach to the arbiter. This would also allow us to simplify the possible state transitions a channel can take while under the breach arbiter's vigilance.

### CODE REVIEW

File: breacharbiter_test.go [hunk 1]
```diff
@@ -0,0 +1,782 @@
...
+	})
+	if err != nil {
+		t.Fatalf("unable to list retributions in db: %v", err)
+	}
+	return count
+}
+
+// testRetributionStore executes a generic test suite for any concrete
+// implementation of the RetributionStore interface.
+func testRetributionStoreAddRemove(frs FailingRetributionStore, t *testing.T) {
```
Comment (Roasbeef) [inline breacharbiter_test.go]: The function name and `godoc` comment are mismatched.

File: breacharbiter_test.go [hunk 2]
```diff
@@ -0,0 +1,782 @@
...
+	// Make sure that a new retribution store is actually emtpy.
+	if count := countRetributions(t, frs); count != 0 {
+		t.Fatalf("expected 0 retributions, found %v", count)
+	}
+
+	// Add all retributions, check that ForAll returns the correct
+	// information, and then remove all retributions.
+	testRetributionStoreAdds(frs, t, false)
```
Comment (Roasbeef) [inline breacharbiter_test.go]: We may want to consider making these test cases into proper [sub-tests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks). As this way, it'll be easy to isolate a particular case for execution within the wider test case.

File: lnd_test.go [hunk 3]
```diff
@@ -3115,6 +3412,10 @@ var testsCases = []*testCase{
 		name: "revoked uncooperative close retribution",
 		test: testRevokedCloseRetribution,
 	},
+	{
+		name: "revoked uncooperative close retribution post breach conf",
+		test: testRevokedCloseRetributionPostBreachConf,
```
Comment (Roasbeef) [inline lnd_test.go]: This set and the one above deem to have a good bit of duplication within their test code, and also of their assertions within the tests. What's the rationale of creating an entirely new test? Why not just roll this new one into the original test case?

File: lnwallet/signdescriptor.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/roasbeef/btcd/btcec"
+	"github.com/roasbeef/btcd/txscript"
+	"github.com/roasbeef/btcd/wire"
+)
+
+var (
+	// ErrTweakOverdose signals a SignDescriptor is invalid because both of its
+	// SingleTweak and DoubleTweak are non-nil.
+	ErrTweakOverdose = errors.New("sign descriptor should only have one tweak")
```
Comment (Roasbeef) [inline lnwallet/signdescriptor.go]: kek 😝

File: lnwallet/witnessgen.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+// WitnessGenerator represents a function which is able to generate the final
+// witness for a particular public key script. This function acts as an
+// abstraction layer, hiding the details of the underlying script.
+type WitnessGenerator func(tx *wire.MsgTx, hc *txscript.TxSigHashes,
+	inputIndex int) ([][]byte, error)
+
+// GenWitnessFunc will return a WitnessGenerator function that an output
+// uses to generate the witness for a sweep transaction.
+func (wt WitnessType) GenWitnessFunc(signer *Signer,
```
Comment (Roasbeef) [inline lnwallet/witnessgen.go]: Very cool usage of Go's ability to add methods to arbitrary types 👍

Comment (cfromknecht) [inline lnwallet/witnessgen.go]: credit goes to @phlip9 on this one :)

File: breacharbiter.go [hunk 6]
```diff
@@ -541,7 +709,8 @@ func (b *breachArbiter) breachObserver(contract *lnwallet.LightningChannel,
 			desc.SigHashes = hc
 			desc.InputIndex = inputIndex
 
-			return lnwallet.CommitSpendNoDelay(b.wallet.Cfg.Signer, &desc, tx)
+			return lnwallet.CommitSpendNoDelay(
+				b.wallet.Cfg.Signer, &desc, tx)
```
Comment (cfromknecht) [inline breacharbiter.go]: I have resolved and simplified the logic for doing so in my future PR for sweeping second layer HTLCs. Talked w/roasbeef and we have decided to leave as is and apply the change in the subsequent PR.

File: lnd_test.go (hunk 3 above)
Comment (cfromknecht) [inline lnd_test.go]: I would like to roll these together, though initially this was done to ensure that we were working with a clean slate, and that i wasn't introducing bugs into the test case itself. After talking w/roasbeef, final verdict was to refactor in 2nd-layer htlc PR, along with more robust refactoring of the unit tests.

File: breacharbiter_test.go (hunk 2 above)
Comment (cfromknecht) [inline breacharbiter_test.go]: Similar to below, we will refactor breach arbiter tests in subsequent PR.

**Comment by cfromknecht** [comment]
@Roasbeef thank you for your review! I have pushed edits fixing your comments regarding: not trimming the `channelsToWatch` slice, reformatting to use the lnd standard for line breaking, test function documentation, and simplifying retribution store logic by directly returning the error from the final subtask.  As discussed, we will postpone the changes related to refactoring of test cases and witness generation function instantiation to a later PR, as some of these changes are already completed/in development. I've updated the body of the PR to note the deference of these tasks.

File: lnd_test.go [hunk 7]
```diff
@@ -1876,35 +1904,295 @@ func testRevokedCloseRetribution(net *networkHarness, t *harnessTest) {
...
+	if err != nil {
+		t.Fatalf("unable to query for alice's channels: %v", err)
+	}
+	if len(aliceChanInfo.Channels) != 0 {
+		t.Fatalf("alice shouldn't have a channel: %v",
+			spew.Sdump(aliceChanInfo.Channels))
+	}
+}
+
+func testRevokedCloseRetributionPostBreachConf(
```
Comment (halseth) [inline lnd_test.go]: Can you add a comment to these two integration tests to explain the two different scenarios they test? :)

Comment (halseth) [inline lnd_test.go]: Or if this is to be refactored then it might not be that important, but because of the duplication atm it is hard to quickly glance where they are different.

File: lnwallet/witnessgen.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+// WitnessType determines how an output's witness will be generated. The
+// default commitmentTimeLock type will generate a witness that will allow
+// spending of a time-locked transaction enforced by CheckSequenceVerify.
+type WitnessType uint16
+
+const (
+	// CommitmentTimeLock is a witness that allows us to spend the output of a
+	// commitment transaction after a relative lock-time lockout.
+	CommitmentTimeLock WitnessType = 0
```
Comment (halseth) [inline lnwallet/witnessgen.go]: Could `iota` be used in this case?

File: utxonursery.go [hunk 9]
```diff
@@ -14,7 +14,7 @@ import (
 	"github.com/lightningnetwork/lnd/chainntnfs"
 	"github.com/lightningnetwork/lnd/channeldb"
 	"github.com/lightningnetwork/lnd/lnwallet"
-	"github.com/roasbeef/btcd/btcec"
+	//"github.com/roasbeef/btcd/btcec"
```
Comment (halseth) [inline utxonursery.go]: remove?

File: breacharbiter.go [hunk 10]
```diff
@@ -462,8 +627,9 @@ func (b *breachArbiter) breachObserver(contract *lnwallet.LightningChannel,
 		//
 		// TODO(roasbeef): also notify utxoNursery, might've had
 		// outbound HTLC's in flight
-		go waitForChanToClose(uint32(closeInfo.SpendingHeight), b.notifier,
-			nil, chanPoint, closeInfo.SpenderTxHash, func() {
+		go waitForChanToClose(uint32(closeInfo.SpendingHeight),
```
Comment (cfromknecht) [inline breacharbiter.go]: In this specific case, the call back is not guaranteed to be executed, so the program could hang indefinitely during a shutdown. Lalou discovered this in the server/peer PR. Down the line it may be useful to add some sort of synchronization to these sorts of methods, so that it can signal the mutex regardless of whether the callback executed.

It's also somewhat unclear in this case that the breach arbiter should necessarily be responsible for this go routine, IIRC @Roasbeef pointed out that maybe the notifier should be responsible? For now the one is unsupervised, but we should find a loving home for it in the future.

File: lnd_test.go (hunk 7 above)
Comment (cfromknecht) [inline lnd_test.go]: Sure thing! We are planning to refactor these into a single test case, but in the meantime I agree that a comment would be helpful for reviewers :)

In the first test, the restart occurs after the breach txn has been added to the mempool (line 1925). In the second test Alice is restarted (line 2177), occurring after the breach txn is confirmed.  Once we've completed some more test refactoring we should be able to step through these cases with more granularity and enumerate the edge cases more concretely, but for now we only have rough control over the failure of the daemon (instead of individual writes, message passing, subsystems, etc)

File: lnwallet/witnessgen.go (hunk 8 above)
Comment (cfromknecht) [inline lnwallet/witnessgen.go]: certainly, great idea!

File: breacharbiter.go [hunk 11]
```diff
@@ -103,15 +152,39 @@ func (b *breachArbiter) Start() error {
...
+	// The state machines generated for these filtered channels can be
+	// discarded, as their fate will be placed in the hands of an
+	// exactRetribution task spawned later.
+	//
+	// NOTE Spawning of the exactRetribution task is intentionally postponed
+	// until after this step in order to ensure that the all breached
+	// channels are reflected as closed in channeldb and consistent with
+	// what is checkpointed by the breach arbiter. Instead of treating the
+	// breached-and-closed and breached-but-still-active channels as
+	// separate sets of channels, we first
```
Comment (cfromknecht) [inline breacharbiter.go]: not sure how this go here haha, thank you

File: breacharbiter.go [hunk 12]
```diff
@@ -103,15 +152,39 @@ func (b *breachArbiter) Start() error {
...
+	// ensure that all breach-but-still-active channels are promoted to
+	// breached-and-closed during restart, allowing us to treat them as a
+	// single set from here on out. This approach also has the added benefit
+	// of minimizing the likelihood that the wrong number of tasks are
+	// spawned per breached channel, and prevents us from being in a
+	// position where
```
Comment (cfromknecht) [inline breacharbiter.go]: same as above

File: breacharbiter.go [hunk 13]
```diff
@@ -1,21 +1,35 @@
...
 	"github.com/roasbeef/btcutil"
 )
 
+// retributionBucket stores retribution state on disk between detecting a
+// contract breach, broadcasting a justice transaction that sweeps the channel,
+// and finally witnessing the justice transaction confirm on the blockchain. It
+// is critical that such state is persisted on disk, so that if our node
+// restarts at any point during the retribution procedure, we can recover and
+// continue from the persisted state.
+var retributionBucket = []byte("ret")
```
Comment (cfromknecht) [inline breacharbiter.go]: I don't have enough knowledge of other bucket names, but we could choose something more distinctive at the very least?

File: breacharbiter.go [hunk 14]
```diff
@@ -556,31 +718,69 @@ func (b *breachArbiter) breachObserver(contract *lnwallet.LightningChannel,
...
+				witnessFunc:    remoteWitness,
 			},
 
+			htlcOutputs: []*breachedOutput{},
+
 			doneChan: make(chan struct{}),
 		}
 
+		// Persist the pending retribution state to disk.
+		if err := b.retributionStore.Add(retInfo); err != nil {
```
Comment (cfromknecht) [inline breacharbiter.go]: Technically yes, a failure of either could cause the breach arbiter to enter a potentially invalid state. The issue here is that we currently don't have a mechanism for distinguishing fatal vs temporary errors. Perhaps the safest behavior in the meantime would be to only shutdown if we continually fail after some number of retries? Hopefully @Roasbeef can shed some wisdom here.

File: breacharbiter.go (hunk 13 above)
Comment (halseth) [inline breacharbiter.go]: If you look at the buckets in `channeldb/channel.go`, the newer one has gotten more distinctive names (e.g. `chanConfigPrefix   = []byte("chan-config")`). I think giving this something simple like `retribution` would do :)

File: breacharbiter.go (hunk 10 above)
Comment (halseth) [inline breacharbiter.go]: I see :) Might make sense to add (partially) this explanation as a comment.

File: lnwallet/witnessgen.go [hunk 15]
```diff
@@ -15,16 +15,16 @@ type WitnessType uint16
 const (
 	// CommitmentTimeLock is a witness that allows us to spend the output of a
 	// commitment transaction after a relative lock-time lockout.
-	CommitmentTimeLock WitnessType = 0
+	CommitmentTimeLock WitnessType = iota
```
Comment (Roasbeef) [inline lnwallet/witnessgen.go]: With this change, it's possible that a later movement or shift int he ordering of the declared `WitnessType` values could result in a breaking database change (and a possible inability to properly sweep outputs). This is due to the fact that the persistence code uses the witness type directly in the on-disk serialization of the contested outputs.  

To avoid a situation like this, we either need to revert this change, or use a distinct mapping with the database code itself. Personally, I have the latter option as it further isolates the daemon's database serialization from a define type in `lnwallet`.

Comment (halseth) [inline lnwallet/witnessgen.go]: Agree that the latter sounds better as it decouples the two. Maybe with an error returned from the database code if it encounters an unknown WitnessType.

File: lnwallet/witnessgen.go (hunk 8 above)
Comment (cfromknecht) [inline lnwallet/witnessgen.go]: Ended up reverting this so that each witness type has a concrete value, this seemed to be the simplest approach to ensuring that ordering didn't corrupt any persistent state. Might try to fully decouple these in my next PR or in the process of packaging up various subsystems in main.

File: breacharbiter.go (hunk 10 above)
Comment (cfromknecht) [inline breacharbiter.go]: Added a comment to document this exception :)

File: breacharbiter.go [hunk 16]
```diff
@@ -103,15 +152,37 @@ func (b *breachArbiter) Start() error {
...
+	// Here we will determine a set of channels that will need to be managed
+	// by the contractObserver.  For each of the open channels read from
+	// disk, we will create a channel state machine that can be used to
+	// watch for any potential channel closures.  We must first exclude any
+	// channel whose retribution process has been initiated, and proceed to
+	// mark them as closed.  The state machines generated for these filtered
+	// channels can be discarded, as their fate will be placed in the hands
+	// of an exactRetribution task spawned later.
+	//
+	// NOTE Spawning of the exactRetribution task is intentionally postponed
```
Comment (halseth) [inline breacharbiter.go]: NOTE:
